pkg/archive/archive.go

```diff
@@ -530,7 +550,8 @@ func createTarFile(path, extractDir string, hdr *tar.Header, reader io.Reader, L
 		}
 
 	case tar.TypeLink:
-		targetPath := filepath.Join(extractDir, hdr.Linkname)
+		lpath := adjustHardLink(hdr.Linkname, hdr.Name)
+		targetPath := filepath.Join(extractDir, lpath)
```

Comment:
I don't agree that the bug is here. I think the bug is in the generation of the `tar` archive, because it doesn't make sense for `Linkname` to be `tmp/a` in this situation (I actually don't like how we do `docker cp` path fudging but that ship sailed 3 years ago).